cmd/dendrite/main.go

```diff
@@ -157,13 +157,14 @@ func main() {
 
 	keyRing := fsAPI.KeyRing()
 
-	userAPI := userapi.NewInternalAPI(processCtx, cfg, cm, &natsInstance, rsAPI, federationClient)
-	asAPI := appservice.NewInternalAPI(processCtx, cfg, &natsInstance, userAPI, rsAPI)
-
 	// The underlying roomserver implementation needs to be able to call the fedsender.
 	// This is different to rsAPI which can be the http client which doesn't need this
 	// dependency. Other components also need updating after their dependencies are up.
 	rsAPI.SetFederationAPI(fsAPI, keyRing)
+
+	userAPI := userapi.NewInternalAPI(processCtx, cfg, cm, &natsInstance, rsAPI, federationClient)
+	asAPI := appservice.NewInternalAPI(processCtx, cfg, &natsInstance, userAPI, rsAPI)
```

Comment:
It would be good to document this ordering dependency somewhere. Maybe both in userapi.NewInternalAPI & appservice.NewInternalAPI?
It would help in the future to determine why this ordering of initialization matters.

The dependency sucks, but our service initialization is already terrible and needs a rework (which we have longer term plans for) so adding in this isn't exactly out of line with the existing usage :face_exhaling:

Reply:
Good idea! And I definitely agree it needs a rework. Committed